Current user conditional removal

AllPurposeName opened the pull request:
I like to poop

AllPurposeName committed:
Added Draper gem for easy decorating.

AllPurposeName committed:
implemented a basic current_user link_to. Create room existence is up next.

AllPurposeName committed:
added rubymine .idea/ to .gitignore

AllPurposeName committed:
fixed a few minor things.

A test was failing due to  since current_user always exists. Other than that there were a couple typos on my part causing other tests to fail.

AllPurposeName committed:
Template for new_room was not being loaded

AllPurposeName committed:
Refactored create_room conditionals.

AllPurposeName committed:
Create room button does not appear while creating room.

AllPurposeName committed:
new_room_form now uses CU#form_link.

This implementation was very easy having created two similar methods prior.
The issue came in mocking the form in the test. Had to include the Draper
helper methods. Line 32 is creating the form to send through to the Current
User.

AllPurposeName committed:
'Close this Topic' no longer conditonally linked to current_user

AllPurposeName committed:
Kyra's most recent PR made irrelevant most of my changes.

I kept a few of them to show off, but I will happily delete
them once I demonstrate what I've implemented and how.
Draper is pretty neat, but so many of the conditionals are in the
nitty gritty of the rails helpers, so when using JavaScript
it doesn't even help.

indiesquidge commented on app/decorators/current_user.rb:
```diff
@@ -0,0 +1,46 @@
+      ""
+    end
+  end
+end
```
You really feel like this file is a better approach than having an if/else in a view? It's much harder to understand and read through. Like what is `h`? Why are there `else` blocks that return empty strings? It's great to see a decorator used, but it just seems like a ton of overhead for something that doesn't need to be optimized.

AllPurposeName replied:
I do feel like it's a better approach. In the view there is one line with only one method which is easy to navigate to. Yeah else can be nil too, don't know why I thought `""` was necessary. h is short for Rails View Helpers such as link_to, form_for, etc. I don't like the voodoo Draper does to gain access to those in the decorator, but I think Rails is already doing enough magic for this to be justifiable.

indiesquidge replied:
I agree it looks cleaner, but at what value? It doesn't matter for the scope of this project, but if this were going to be something that had to be maintained, it would be a much longer on boarding process to get someone familiar with since it's highly configuration over convention. Plus it was engineered before there was even a problem. I'm fine with it here since I think it's cool you used it and its a great tool to abstract pieces of code, but just something to consider in the future.

AllPurposeName replied:
I think conditionals in the view are neither conventional nor OO and that's the problem I'm trying to solve with this branch. I would use a PORO, but then I would have to pass the view context into that PORO anytime it needed to generate a link. I think getting a current_user functional early is good for the same reason that anything which reduces your technical debt is important early. Anytime we need a conditional in the view from here on it would be a single test in the decorator spec, a single method in the current user, and a single method in the view. I think that's so much better than something like `<%= if current_user && @message.user == current_user || current_user.admin? || current_user.moderator? %>` Which from the re-pivot I had A LOT of issues with.

Onboarding would be an issue, which makes me interested to see what Horace will say in our assessment because he worked on a giant product which probably had much more bewildering things new employees had to deal with.

indiesquidge commented on app/views/rooms/show.html.erb:
```diff
@@ -5,59 +5,59 @@
-  </div> <!-- row -->
-</div> <!-- container -->
+    </div> <!-- row -->
+  </div> <!-- container -->
```
Tabbing is off in this file. Run `gg=G` :)
Orrrrrr....
You can use this nice vim key binding that Josh Mejia showed me to auto-indent your file based on the file type:

``` viml
" Use leader-i to properly indent lines depending on filetype
map <Leader>i mzgg=G`z<cr>
```

With this, running leader-i is essentially like running `gg=G`, but it will save your cursor position in a vim register so you don't lose track of where you are.

AllPurposeName replied:
Oh yeah I remember when Mejia showed us this I wasn't using vim yet. Dope thanks for the resource.

indiesquidge commented on app/views/shared/_new_room_form.html.erb:
```diff
@@ -0,0 +1,12 @@
+      <% end %>
+    </div>
+  </div>
+</div>
```
How come we are adding back in this file? It is not reachable by a user. If you just go to the route, it also displays two forms, ha!
![screen shot 2015-06-21 at 11 34 06 am](https://cloud.githubusercontent.com/assets/3409645/8272668/7996b1f4-1809-11e5-93c6-ca0112609e47.png)

AllPurposeName replied:
Yeah I think I said in slack that after playing with the app having an entire view for creating a room is unnecessary. I only wanted to keep this one in for the PR so I could show off how to use Draper to dynamically set the f.submit in a form_for.

AllPurposeName replied:
I have already removed it, just waiting for other changes before I push.

AllPurposeName committed:
Took care of Current User redundancies and spacing.

I switched the instances of  to  for better readability
within the current_user decorator. The various spacing mistakes were also
taken care of. Removed new_room for good.
